packages/x-data-grid/src/components/cell/GridCell.tsx

```diff
@@ -183,33 +190,29 @@ const GridCell = React.forwardRef<HTMLDivElement, GridCellProps>(function GridCe
     ...other
   } = props;
 
-  const apiRef = useGridApiContext();
+  const apiRef = useGridPrivateApiContext();
   const rootProps = useGridRootProps();
   const isRtl = useRtl();
 
   const field = column.field;
 
-  const cellParams = useGridSelector(
-    apiRef,
-    () => {
-      // This is required because `.getCellParams` tries to get the `state.rows.tree` entry
-      // associated with `rowId`/`fieldId`, but this selector runs after the state has been
-      // updated, while `rowId`/`fieldId` reference an entry in the old state.
-      const row = apiRef.current.getRow(rowId);
-      if (!row) {
-        return EMPTY_CELL_PARAMS;
-      }
-
-      const result = apiRef.current.getCellParams<any, any, any, GridTreeNodeWithRender>(
-        rowId,
-        field,
-      );
-      result.api = apiRef.current;
-      return result;
-    },
-    objectShallowCompare,
+  const cellParams = apiRef.current.getCellParams<any, any, any, GridTreeNodeWithRender>(
```

Comment:
That's actually one of the core changes that is a step in the right direction imho, tried to explain it briefly here: https://github.com/mui/mui-x/pull/15627#issuecomment-2506490646

The current behaviour casts an incredibly wide net to cell-level reactivity that opens it up to race conditions, and prevents from carefully managing state propagation.

Should cell valueGetter and valueFormatter of each cell in the datagrid (renderecontext) re-evaluate whenever any part of grid state changes? (focus, tabindex, rendercontext, etc.?). The currently do, but I would argue they shouldn't. They should only be re-evaluated if the row changes (through edits or row updates).

Imagine if you define a valueFormatter that formats the cell value based on external state. Changing that external state wouldn't update the rendered cell value, but changing the focus or scrolling  a bit would suddenly trigger the cell to re-render with a new value. What?!

Want to understand why the cell re-rendered? Most of the cases, the answer is that getCellParams changed. But why did they change? Good luck understanding that.

So, ideally:
* Row data is provided by row (or editing api)
* Rest of the conditions why the cell should re-render independent of the row should be carefully managed based on atomic selectors. As you can see, the scope currently is not that wide in order to make the tests pass. This should result in both more optimised code as well as more understandable state flow within the grid.

Reply:
Okay, this explains why `tabIndex` and `hasFocus` are overridden with the values using `useGridSelector`.

@romgrk Does this make sense to you?

Reply:
> Imagine if you define a valueFormatter that formats the cell value based on external state. Changing that external state wouldn't update the rendered cell value, but changing the focus or scrolling a bit would suddenly trigger the cell to re-render with a new value. What?!

The solution to that would be the approach that we've been introducing of adding some props to the grid's store reactivity model. In this case, `valueFormatter` changes should trigger a store update.

> @romgrk Does this make sense to you?

No, it still doesn't. AFAICT, we would be losing reactivity for the `cellParams` values other than the ones overriden with `useGridSelector` values. And I see some of them for which it would be problematic, e.g. `cellMode` or `isEditable`.

> Want to understand why the cell re-rendered? Most of the cases, the answer is that getCellParams changed. But why did they change? Good luck understanding that.

`getCellParams` has been problematic for a long time (perf-wise and logic-wise), I haven't been able to get rid of it because it would be a breaking change, some of the API uses `GridCellParams`. I think if we want to fix everything cleanly, we need to decompose that function into more logical parts than "here's every bit of cell data". But that would prevent cherry-picking this PR in v7.

Thoughts?

Reply:
I don't think that's quite true that reactivity is lost. A bunch of tests would be failing if that were the case, and cell editing wouldn't work. Any `colDef` update propagates through the row as it is already. Whereas `tabIndex` and `focus` live in unrelated store to columns/rows.

Complete decomposition and granular cell-level reactivity would be a much bigger rewrite. My quick assessment currently is that it's not worth it and would needlessly complicate the code, as there's too many interrelated layers. One would need to ask what are the specific use cases where this would bring meaningful performance gains – I'm personally not seeing it. Maybe editing in theory, but since editing apis expose the whole row for various reasons, that's also not really the case.

Reply:
https://github.com/mui/mui-x/blob/521c696fd095520fb4fdde75b5b42875eb57d4a5/packages/x-data-grid/src/components/GridRow.tsx#L338

Reply:
> A bunch of tests would be failing if that were the case

But I don't think we have 100% test coverage, I think we're in the 80s or 90s.

> I don't think that's quite true that reactivity is lost.

But there is still nothing that explicitly binds cell reactivity to those values; the binding is more like a coincidental side-effect that might break under edge-cases or unrelated refactors. If you ask me why `tabIndex` works, it's because it's bound through `useGridSelector`. If you ask me why `isEditable` works: I don't know. And I feel uncomfortable merging code that works for unknown reasons.

> Complete decomposition and granular cell-level reactivity would be a much bigger rewrite.

It doesn't feel like a big change to me, splitting `getCellParams` into a few small functions is all it would take. I'll take a look at the code again tomorrow to see if I can do something v7-compatible.

Reply:
I agree with the rationale that it shouldn't work due to side-effects. However, `isEditable` is not the case – every input to `isEditable` already triggers a re-render through the row->cell tree. It's not a side-effect by any means, but logical flow of state. Adding a selector as well just introduces extraneous work for no benefit. It will simply trigger a re-render from both the Cell within and Row as well.

On a second look, the only thing I actually think needs extracting out is `cellMode`, which does probably work due to side-effects.

The mental model to me is:
1. Everything in the row -> cell tree that depends directly on the row / cell store shouldn't need additional selectors, as the mere existence of a rendered cell means it's already reactive to these slices of the store.
2. Additional global state that affects how the particular cell is rendered (tabIndex, focus, cellMode) needs selectors.

Does that make sense to you or am I thinking completely wrong here?

Reply:
Or actually, `cellMode` is not even a sideEffect – it depends on the same slice of the state that is already selected in the `GridRow`:
https://github.com/mui/mui-x/blob/2700d7e093c64873ac8fd5af5b58e22fefa11f28/packages/x-data-grid/src/components/GridRow.tsx#L346

`editCellState` prop:
https://github.com/mui/mui-x/blob/2700d7e093c64873ac8fd5af5b58e22fefa11f28/packages/x-data-grid/src/components/GridRow.tsx#L322

`getCellMode` prop depends on the same slice of the state:
https://github.com/mui/mui-x/blob/2700d7e093c64873ac8fd5af5b58e22fefa11f28/packages/x-data-grid/src/hooks/features/editing/useGridCellEditing.ts#L254-L261

So, there are two options:
1. `cellMode` should be a prop of a cell
2. `editRowsState` selector should be removed from the row, and there should be a `editCellState` selector withing `GridCell`. On first look, I don't see why the whole row needs to re-render in response to that, since it's only used to pass a prop to cells, so it could be handled within cells themselves.

Reply:
As long as it's an explicit reactive binding (through a prop or through a selector) I'm happy with that, but option 2 sounds better. I didn't touch the editing code when I did the reactivity refactor so there's leftover stuff from the old "reactivity" architecture (re-render everything anytime anything changes).

Once the editing state is split into an `editCellState` selector, all that is left to do would be to split `value` and `formattedValue` logic, and we could get rid of `getCellParams` in `GridCell`. The other values (`id, field, row, rowNode, colDef`) are all available through props.

If you want to go ahead with that please do, otherwise I'll see if I can complete this tomorrow. Btw we appreciate all the PRs you submitted, we're a bit short on time with v8 preparation and we've been slow to review & merge them, but they're all great.

Reply:
Fixed it up now. Seems to work well. There was one thing that seems to have worked due to side-effects (already before this PR) in the GridRow, moved it to a discrete selector:
https://github.com/mui/mui-x/blob/33cc4c95a262da6953510b339c4916fa3d565873/packages/x-data-grid/src/components/GridRow.tsx#L127

There are some additional re-renderings that can be fixed when stopping editing that should speed up editing in large grids, but that should be done when other PRs have been merged around memoization + GridRoot + `forwardRef` patch.

I'm a bit reluctant to drop `getCellParams` altogether in `GridCell` as it opens up an easy divergence between the public API and internal logic. I refactored it a bit to make sure we don't do any unnecessary operations there though, but still keep a solid contract. Unless we need to extract `value` and `formattedValue` logic into selector hooks, I don't really see the point either. And I don't currently see a reason why we would need to use hooks there. If we do, we should come up with a test case that it solves, otherwise it's probably just redundant operations.